src/lib.rs

```diff
@@ -208,13 +208,45 @@ where
   _p_c2: PhantomData<C2>,
 }
 
-impl<G1, G2, C1, C2> SimpleDigestible for PublicParams<G1, G2, C1, C2>
+impl<G1, G2, C1, C2> Digestible for PublicParams<G1, G2, C1, C2>
 where
   G1: Group<Base = <G2 as Group>::Scalar>,
   G2: Group<Base = <G1 as Group>::Scalar>,
   C1: StepCircuit<G1::Scalar>,
   C2: StepCircuit<G2::Scalar>,
+  G1::Scalar: Serialize,
 {
+  fn write_digest<W: Sized + io::Write>(&self, byte_sink: &mut W) -> Result<(), io::Error> {
+    self.digest.write_digest(byte_sink)
+  }
+
+  fn to_bytes(&self) -> Result<Vec<u8>, io::Error> {
```

Comment:
Perhaps we could implement a quick proc macro for this

Reply:
`to_bytes` is what creates a digest from a raw object `T`. `write_digestible_bytes` is how `T` should be represented when creating the digest of a larger object `S` can contains `T`. For `R1CSShape` and `PublicParams` we break the invariant because when computing the digest of `PublicParams` that contains `R1CSShape`s, we simply want to absorb `shape.digest` instead of rehashing all of `shape` again. Thus the implementation of `write_digestible_bytes` for `R1CSShape` should just write its digest and nothing else.

Reply:
But now for all structures that use the digests of their constituent fields (rather than the serialization bytes of their constituent fields), I have two ways of getting to a digest for the top-level structure: hashing the output of `to_bytes`, and asking `write_digestable_bytes` to write the digest to a byte sink. 

For some structures this is the same. For some this is not. This ambiguity is bound to end in tears.